File: src/nexmap/blueprint.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

// --- Resolution tables ---

var sizeRanges = map[string][2]int{
	"small":  {256, 384},
	"medium": {384, 640},
	"large":  {640, 896},
	"huge":   {896, 1024},
}

var elevationRanges = map[string][2]int{
	"ground": {0, 0},
	"low":    {-128, -64},
	"mid":    {64, 192},
	"high":   {256, 384},
	"varied": {-128, 384},
}

var ceilingMap = map[string]int{
	"low":       128,
	"normal":    192,
	"tall":      256,
	"cavernous": 320,
}

var connectionWidths = map[string]int{
	"narrow": 64,
	"normal": 96,
	"wide":   160,
}

var hazardCoverage = map[string]float64{
	"small":  0.15,
	"medium": 0.30,
	"large":  0.50,
}

var hazardTextures = map[string]string{
	"lava":  "*lava1",
	"water": "*04water1",
	"slime": "*slime",
}

// --- Raw JSON types ---

type rawBlueprint struct {
	Version     int              `json:"version"`
	Meta        rawMeta          `json:"meta"`
	Rooms       []rawRoom        `json:"rooms"`
	Connections []rawConnection  `json:"connections"`
	Flow        *rawFlow         `json:"flow,omitempty"`
	Textures    map[string]string `json:"textures,omitempty"`
}

type rawMeta struct {
	Name        string          `json:"name"`
	Author      string          `json:"author,omitempty"`
	Description string          `json:"description,omitempty"`
	Theme       string          `json:"theme"`
	Style       string          `json:"style,omitempty"`    // source map style (e.g. "dm4")
	CustomTheme string          `json:"custom_theme,omitempty"`
	Gamemode    string          `json:"gamemode"`
	PlayerCount *rawPlayerCount `json:"player_count,omitempty"`
	ArenaSize   int             `json:"arena_size,omitempty"`
	DesignNotes string          `json:"design_notes,omitempty"`
}

type rawPlayerCount struct {
	Min int `json:"min"`
	Max int `json:"max"`
}

type rawRoom struct {
	ID          string          `json:"id"`
	Role        string          `json:"role"`
	Label       string          `json:"label,omitempty"`
	Size        string          `json:"size,omitempty"`
	Shape       string          `json:"shape,omitempty"`
	Elevation   string          `json:"elevation,omitempty"`
	Ceiling     string          `json:"ceiling,omitempty"`
	Environment string          `json:"environment,omitempty"` // building, cave, outdoor, hallway
	Details     []string        `json:"details,omitempty"`     // pillars, platform, light_recesses, wall_trim, crates, step_down
	Hazard      *rawHazard      `json:"hazard,omitempty"`
	Items       []rawItem       `json:"items,omitempty"`
	Tags        []string        `json:"tags,omitempty"`
}

type rawHazard struct {
	Type     string `json:"type"`
	Coverage string `json:"coverage,omitempty"`
}

type rawItem struct {
	Classname string `json:"classname"`
	Placement string `json:"placement,omitempty"`
	Intent    string `json:"intent,omitempty"`
	Count     int    `json:"count,omitempty"`
}

type rawConnection struct {
	From          string `json:"from"`
	To            string `json:"to"`
	Type          string `json:"type"`
	Bidirectional *bool  `json:"bidirectional,omitempty"`
	Width         string `json:"width,omitempty"`
	Intent        string `json:"intent,omitempty"`
}

type rawFlow struct {
	PrimaryLoop    []string     `json:"primary_loop,omitempty"`
	SecondaryLoops [][]string   `json:"secondary_loops,omitempty"`
	ChokePoints    []string     `json:"choke_points,omitempty"`
	Sightlines     []rawSight   `json:"sightlines,omitempty"`
}

type rawSight struct {
	From        string `json:"from"`
	To          string `json:"to"`
	Description string `json:"description,omitempty"`
}

// --- Resolved types ---

type ResolvedItem struct {
	Classname string
	Placement string
	Intent    string
	Count     int
}

type ResolvedHazard struct {
	Type     string
	Texture  string
	Coverage float64
}

type ResolvedRoom struct {
	ID           string
	Role         string
	Label        string
	SizeMin      int
	SizeMax      int
	Shape        string
	ElevationMin int
	ElevationMax int
	CeilingHeight int
	Environment  string   // building, cave, outdoor, hallway
	Details      []Detail // architectural details to place
	Hazard       *ResolvedHazard
	Items        []ResolvedItem
	Tags         []string
}

type ResolvedConnection struct {
	FromID        string
	ToID          string
	Type          string
	Bidirectional bool
	Width         int
	Intent        string
}

type ResolvedBlueprint struct {
	Version       int
	Name          string
	Author        string
	Description   string
	Theme         string
	Style         string // source map style (e.g. "dm4")
	Gamemode      string
	PlayerMin     int
	PlayerMax     int
	ArenaSize     int
	DesignNotes   string
	Rooms         []ResolvedRoom
	Connections   []ResolvedConnection
	TextureOverrides map[string]string
}

func (bp *ResolvedBlueprint) RoomByID(id string) *ResolvedRoom {
	for i := range bp.Rooms {
		if bp.Rooms[i].ID == id {
			return &bp.Rooms[i]
		}
	}
	return nil
}

// --- Validation ---

func validateBlueprint(raw *rawBlueprint) []string {
	var errs []string

	if raw.Version != 1 {
		errs = append(errs, "version must be 1")
	}
	if raw.Meta.Name == "" {
		errs = append(errs, "meta.name is required")
	}
	if raw.Meta.Theme == "" {
		errs = append(errs, "meta.theme is required")
	}
	if raw.Meta.Gamemode == "" {
		errs = append(errs, "meta.gamemode is required")
	}
	if len(raw.Rooms) < 2 {
		errs = append(errs, "need at least 2 rooms")
	}

	ids := map[string]bool{}
	for i, r := range raw.Rooms {
		if r.ID == "" {
			errs = append(errs, fmt.Sprintf("rooms[%d] missing id", i))
		} else if ids[r.ID] {
			errs = append(errs, fmt.Sprintf("duplicate room id: %s", r.ID))
		}
		ids[r.ID] = true
		if r.Role == "" {
			errs = append(errs, fmt.Sprintf("rooms[%d] missing role", i))
		}
	}

	if len(raw.Connections) < 1 {
		errs = append(errs, "need at least 1 connection")
	}

	// Connection reference check.
	for i, c := range raw.Connections {
		if !ids[c.From] {
			errs = append(errs, fmt.Sprintf("connections[%d].from references unknown room '%s'", i, c.From))
		}
		if !ids[c.To] {
			errs = append(errs, fmt.Sprintf("connections[%d].to references unknown room '%s'", i, c.To))
		}
	}

	// BFS connectivity.
	if len(errs) == 0 {
		adj := map[string][]string{}
		for _, id := range raw.Rooms {
			adj[id.ID] = nil
		}
		for _, c := range raw.Connections {
			adj[c.From] = append(adj[c.From], c.To)
			bidir := c.Bidirectional == nil || *c.Bidirectional
			if bidir {
				adj[c.To] = append(adj[c.To], c.From)
			}
		}
		visited := map[string]bool{}
		queue := []string{raw.Rooms[0].ID}
		for len(queue) > 0 {
			node := queue[0]
			queue = queue[1:]
			if visited[node] {
				continue
			}
			visited[node] = true
			queue = append(queue, adj[node]...)
		}
		var unreachable []string
		for _, r := range raw.Rooms {
			if !visited[r.ID] {
				unreachable = append(unreachable, r.ID)
			}
		}
		if len(unreachable) > 0 {
			errs = append(errs, fmt.Sprintf("unreachable rooms (disconnected graph): %v", unreachable))
		}
	}

	return errs
}

// --- Resolution ---

func resolveBlueprint(raw *rawBlueprint) *ResolvedBlueprint {
	meta := raw.Meta

	playerMin, playerMax := 2, 8
	if meta.PlayerCount != nil {
		if meta.PlayerCount.Min > 0 {
			playerMin = meta.PlayerCount.Min
		}
		if meta.PlayerCount.Max > 0 {
			playerMax = meta.PlayerCount.Max
		}
	}

	arenaSize := 3072
	if meta.ArenaSize > 0 {
		arenaSize = meta.ArenaSize
	}

	theme := meta.Theme
	if theme == "custom" && meta.CustomTheme != "" {
		theme = meta.CustomTheme
	}

	bp := &ResolvedBlueprint{
		Version:     raw.Version,
		Name:        meta.Name,
		Author:      orDefault(meta.Author, "NexMap LLM"),
		Description: meta.Description,
		Theme:       theme,
		Style:       meta.Style,
		Gamemode:    meta.Gamemode,
		PlayerMin:   playerMin,
		PlayerMax:   playerMax,
		ArenaSize:   arenaSize,
		DesignNotes: meta.DesignNotes,
		TextureOverrides: raw.Textures,
	}

	for _, rr := range raw.Rooms {
		bp.Rooms = append(bp.Rooms, resolveRoom(rr))
	}
	for _, rc := range raw.Connections {
		bp.Connections = append(bp.Connections, resolveConnection(rc))
	}

	return bp
}

func resolveRoom(r rawRoom) ResolvedRoom {
	sz := orDefault(r.Size, "medium")
	rng, ok := sizeRanges[sz]
	if !ok {
		rng = sizeRanges["medium"]
	}

	elev := orDefault(r.Elevation, "ground")
	er, ok := elevationRanges[elev]
	if !ok {
		er = elevationRanges["ground"]
	}

	ceil := orDefault(r.Ceiling, "normal")
	ch, ok := ceilingMap[ceil]
	if !ok {
		ch = ceilingMap["normal"]
	}

	var hazard *ResolvedHazard
	if r.Hazard != nil {
		cov := 0.30
		if v, ok := hazardCoverage[orDefault(r.Hazard.Coverage, "medium")]; ok {
			cov = v
		}
		hazard = &ResolvedHazard{
			Type:     r.Hazard.Type,
			Texture:  hazardTextures[r.Hazard.Type],
			Coverage: cov,
		}
	}

	var items []ResolvedItem
	for _, ri := range r.Items {
		count := ri.Count
		if count < 1 {
			count = 1
		}
		items = append(items, ResolvedItem{
			Classname: ri.Classname,
			Placement: orDefault(ri.Placement, "any"),
			Intent:    ri.Intent,
			Count:     count,
		})
	}

	var details []Detail
	for _, d := range r.Details {
		details = append(details, Detail(d))
	}

	return ResolvedRoom{
		ID:           r.ID,
		Role:         r.Role,
		Label:        orDefault(r.Label, r.ID),
		SizeMin:      rng[0],
		SizeMax:      rng[1],
		Shape:        orDefault(r.Shape, "any"),
		ElevationMin: er[0],
		ElevationMax: er[1],
		CeilingHeight: ch,
		Environment:  orDefault(r.Environment, "building"),
		Details:      details,
		Hazard:       hazard,
		Items:        items,
		Tags:         r.Tags,
	}
}

func resolveConnection(c rawConnection) ResolvedConnection {
	bidir := true
	if c.Bidirectional != nil {
		bidir = *c.Bidirectional
	}
	w := 96
	if v, ok := connectionWidths[orDefault(c.Width, "normal")]; ok {
		w = v
	}
	return ResolvedConnection{
		FromID:        c.From,
		ToID:          c.To,
		Type:          c.Type,
		Bidirectional: bidir,
		Width:         w,
		Intent:        c.Intent,
	}
}

// --- Public API ---

func LoadBlueprint(path string) (*ResolvedBlueprint, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("reading blueprint: %w", err)
	}

	var raw rawBlueprint
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("parsing blueprint JSON: %w", err)
	}

	if errs := validateBlueprint(&raw); len(errs) > 0 {
		return nil, fmt.Errorf("blueprint validation failed:\n  - %s", strings.Join(errs, "\n  - "))
	}

	return resolveBlueprint(&raw), nil
}

func orDefault(s, def string) string {
	if s == "" {
		return def
	}
	return s
}

File: src/mapgen/blueprint.py
"""Load, validate, and resolve NexMap blueprints.

A blueprint is the LLM-authored JSON that describes *what* a map should be.
This module validates it against the schema, checks graph connectivity,
and resolves semantic fields (e.g. size='large') into concrete Quake-unit
ranges that the layout compiler can consume.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path

# ---------------------------------------------------------------------------
# Semantic -> concrete resolution tables
# ---------------------------------------------------------------------------

# Room footprint ranges (min_side, max_side) in Quake units.
SIZE_RANGES: dict[str, tuple[int, int]] = {
    "small":  (256, 384),
    "medium": (384, 640),
    "large":  (640, 896),
    "huge":   (896, 1024),
}

# Floor elevation ranges (z0_min, z0_max).
ELEVATION_RANGES: dict[str, tuple[int, int]] = {
    "ground": (0, 0),
    "low":    (-128, -64),
    "mid":    (64, 192),
    "high":   (256, 384),
    "varied": (-128, 384),
}

# Floor-to-ceiling heights.
CEILING_HEIGHTS: dict[str, int] = {
    "low":       128,
    "normal":    192,
    "tall":      256,
    "cavernous": 320,
}

# Connection widths in Quake units.
CONNECTION_WIDTHS: dict[str, int] = {
    "narrow": 64,
    "normal": 96,
    "wide":   160,
}

# Hazard pool coverage as fraction of room floor area.
HAZARD_COVERAGE: dict[str, float] = {
    "small":  0.15,
    "medium": 0.30,
    "large":  0.50,
}

# Hazard type -> texture.
HAZARD_TEXTURES: dict[str, str] = {
    "lava":  "*lava1",
    "water": "*04water1",
    "slime": "*slime",
}

SCHEMA_PATH = Path(__file__).parent / "blueprint_schema.json"


# ---------------------------------------------------------------------------
# Resolved dataclasses — concrete values ready for the compiler
# ---------------------------------------------------------------------------

@dataclass
class ResolvedItem:
    classname: str
    placement: str  # center, corner, wall, elevated, hidden, near_hazard, any
    intent: str
    count: int


@dataclass
class ResolvedHazard:
    type: str          # lava, water, slime
    texture: str       # *lava1, etc.
    coverage: float    # 0.0-1.0 fraction of floor area


@dataclass
class ResolvedRoom:
    id: str
    role: str
    label: str
    size_min: int      # minimum side length (Quake units)
    size_max: int      # maximum side length
    shape: str         # square, wide, tall_narrow, l_shape, any
    elevation_min: int # z0 range low
    elevation_max: int # z0 range high
    ceiling_height: int
    hazard: ResolvedHazard | None
    items: list[ResolvedItem]
    tags: list[str]


@dataclass
class ResolvedConnection:
    from_id: str
    to_id: str
    type: str           # corridor, wide_opening, stairs, drop, teleporter, door
    bidirectional: bool
    width: int          # Quake units
    intent: str


@dataclass
class ResolvedFlow:
    primary_loop: list[str]
    secondary_loops: list[list[str]]
    choke_points: list[str]
    sightlines: list[dict[str, str]]


@dataclass
class ResolvedBlueprint:
    """Fully resolved blueprint ready for the layout compiler."""
    version: int
    name: str
    author: str
    description: str
    theme: str
    gamemode: str
    player_count_min: int
    player_count_max: int
    arena_size: int
    design_notes: str
    rooms: list[ResolvedRoom]
    connections: list[ResolvedConnection]
    flow: ResolvedFlow | None
    textures: dict[str, str]

    def room_by_id(self, room_id: str) -> ResolvedRoom | None:
        for r in self.rooms:
            if r.id == room_id:
                return r
        return None

    def room_ids(self) -> set[str]:
        return {r.id for r in self.rooms}

    def adjacency(self) -> dict[str, list[str]]:
        """Build adjacency list from connections."""
        adj: dict[str, list[str]] = {r.id: [] for r in self.rooms}
        for c in self.connections:
            adj[c.from_id].append(c.to_id)
            if c.bidirectional:
                adj[c.to_id].append(c.from_id)
        return adj


# ---------------------------------------------------------------------------
# Validation
# ---------------------------------------------------------------------------

class BlueprintError(Exception):
    """Raised when a blueprint fails validation."""


def _validate_structure(data: dict) -> list[str]:
    """Basic structural checks (no jsonschema dependency required)."""
    errors = []

    if data.get("version") != 1:
        errors.append("version must be 1")

    meta = data.get("meta")
    if not isinstance(meta, dict):
        errors.append("missing 'meta' object")
    else:
        for key in ("name", "theme", "gamemode"):
            if not meta.get(key):
                errors.append(f"meta.{key} is required")

    rooms = data.get("rooms")
    if not isinstance(rooms, list) or len(rooms) < 2:
        errors.append("need at least 2 rooms")
    else:
        ids = set()
        for i, room in enumerate(rooms):
            rid = room.get("id", "")
            if not rid:
                errors.append(f"rooms[{i}] missing id")
            elif rid in ids:
                errors.append(f"duplicate room id: {rid}")
            ids.add(rid)
            if not room.get("role"):
                errors.append(f"rooms[{i}] missing role")

    conns = data.get("connections")
    if not isinstance(conns, list) or len(conns) < 1:
        errors.append("need at least 1 connection")

    return errors


def _validate_graph(data: dict) -> list[str]:
    """Check that connections reference valid rooms and the graph is connected."""
    errors = []
    room_ids = {r["id"] for r in data.get("rooms", [])}

    for i, conn in enumerate(data.get("connections", [])):
        for end in ("from", "to"):
            rid = conn.get(end, "")
            if rid not in room_ids:
                errors.append(
                    f"connections[{i}].{end} references unknown room '{rid}'"
                )

    # BFS connectivity check (bidirectional edges by default).
    adj: dict[str, list[str]] = {rid: [] for rid in room_ids}
    for conn in data.get("connections", []):
        f, t = conn.get("from", ""), conn.get("to", "")
        if f in room_ids and t in room_ids:
            adj[f].append(t)
            if conn.get("bidirectional", True):
                adj[t].append(f)

    if room_ids:
        start = next(iter(room_ids))
        visited: set[str] = set()
        queue = [start]
        while queue:
            node = queue.pop(0)
            if node in visited:
                continue
            visited.add(node)
            queue.extend(adj.get(node, []))

        unreachable = room_ids - visited
        if unreachable:
            errors.append(
                f"unreachable rooms (disconnected graph): {sorted(unreachable)}"
            )

    # Validate flow references if present.
    flow = data.get("flow")
    if flow:
        for loop_name in ("primary_loop", "secondary_loops"):
            loops = flow.get(loop_name, [])
            if loop_name == "primary_loop":
                loops = [loops] if loops else []
            for loop in loops:
                for rid in loop:
                    if rid not in room_ids:
                        errors.append(
                            f"flow.{loop_name} references unknown room '{rid}'"
                        )

    return errors


def validate(data: dict) -> list[str]:
    """Validate a raw blueprint dict. Returns list of error strings (empty = valid)."""
    errors = _validate_structure(data)
    if not errors:
        errors.extend(_validate_graph(data))
    return errors


# ---------------------------------------------------------------------------
# Resolution — semantic values -> concrete numbers
# ---------------------------------------------------------------------------

def _resolve_room(raw: dict) -> ResolvedRoom:
    size = raw.get("size", "medium")
    size_min, size_max = SIZE_RANGES.get(size, SIZE_RANGES["medium"])

    elev = raw.get("elevation", "ground")
    elev_min, elev_max = ELEVATION_RANGES.get(elev, ELEVATION_RANGES["ground"])

    ceiling = raw.get("ceiling", "normal")
    ceiling_h = CEILING_HEIGHTS.get(ceiling, CEILING_HEIGHTS["normal"])

    hazard = None
    raw_hazard = raw.get("hazard")
    if raw_hazard:
        htype = raw_hazard["type"]
        coverage = HAZARD_COVERAGE.get(
            raw_hazard.get("coverage", "medium"), 0.30
        )
        hazard = ResolvedHazard(
            type=htype,
            texture=HAZARD_TEXTURES[htype],
            coverage=coverage,
        )

    items = []
    for raw_item in raw.get("items", []):
        items.append(ResolvedItem(
            classname=raw_item["classname"],
            placement=raw_item.get("placement", "any"),
            intent=raw_item.get("intent", ""),
            count=raw_item.get("count", 1),
        ))

    return ResolvedRoom(
        id=raw["id"],
        role=raw["role"],
        label=raw.get("label", raw["id"]),
        size_min=size_min,
        size_max=size_max,
        shape=raw.get("shape", "any"),
        elevation_min=elev_min,
        elevation_max=elev_max,
        ceiling_height=ceiling_h,
        hazard=hazard,
        items=items,
        tags=raw.get("tags", []),
    )


def _resolve_connection(raw: dict) -> ResolvedConnection:
    width_key = raw.get("width", "normal")
    return ResolvedConnection(
        from_id=raw["from"],
        to_id=raw["to"],
        type=raw["type"],
        bidirectional=raw.get("bidirectional", True),
        width=CONNECTION_WIDTHS.get(width_key, 96),
        intent=raw.get("intent", ""),
    )


def _resolve_flow(raw: dict | None) -> ResolvedFlow | None:
    if not raw:
        return None
    return ResolvedFlow(
        primary_loop=raw.get("primary_loop", []),
        secondary_loops=raw.get("secondary_loops", []),
        choke_points=raw.get("choke_points", []),
        sightlines=raw.get("sightlines", []),
    )


def resolve(data: dict) -> ResolvedBlueprint:
    """Resolve a validated blueprint dict into concrete compiler-ready values."""
    meta = data["meta"]
    pc = meta.get("player_count", {})

    return ResolvedBlueprint(
        version=data["version"],
        name=meta["name"],
        author=meta.get("author", "NexMap LLM"),
        description=meta.get("description", ""),
        theme=meta.get("custom_theme", meta["theme"])
            if meta["theme"] == "custom" else meta["theme"],
        gamemode=meta["gamemode"],
        player_count_min=pc.get("min", 2),
        player_count_max=pc.get("max", 8),
        arena_size=meta.get("arena_size", 3072),
        design_notes=meta.get("design_notes", ""),
        rooms=[_resolve_room(r) for r in data["rooms"]],
        connections=[_resolve_connection(c) for c in data["connections"]],
        flow=_resolve_flow(data.get("flow")),
        textures=data.get("textures", {}),
    )


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def load_blueprint(path: str | Path) -> ResolvedBlueprint:
    """Load a blueprint JSON file, validate, and resolve to concrete values.

    Raises BlueprintError on validation failure.
    """
    path = Path(path)
    with open(path) as f:
        data = json.load(f)

    errors = validate(data)
    if errors:
        raise BlueprintError(
            f"Blueprint validation failed ({path}):\n"
            + "\n".join(f"  - {e}" for e in errors)
        )

    return resolve(data)


def load_blueprint_str(json_str: str) -> ResolvedBlueprint:
    """Load a blueprint from a JSON string (e.g. direct LLM output)."""
    data = json.loads(json_str)
    errors = validate(data)
    if errors:
        raise BlueprintError(
            "Blueprint validation failed:\n"
            + "\n".join(f"  - {e}" for e in errors)
        )
    return resolve(data)
